### PR TITLE
update normalisation & tests

### PR DESCRIPTION
**Pull request by divyasharma-arup**
Update functionality to normalise density and plot the spatial difference in inputs vs. outputs.

### CODE REVIEW

File: pam/samplers/tour.py [hunk 1]
```diff
@@ -135,9 +151,10 @@ def threshold_sample(self):
         d_threshold = self.distribution[self.distribution.index.isin(d_list)]
         
         if len(d_threshold) == 0:
-            raise UserWarning('No destinations within this threshold value')
-
-        return random.choices(list(d_threshold.index), weights=list(d_threshold[self.frequency]), k=1)[0]        
+            warnings.warn('No destinations within this threshold value, change threshold')
```
Comment (divyasharma-arup) [inline pam/samplers/tour.py:154]: turns out `raise UserWarning` didn't actually raise a warning.

File: pam/samplers/tour.py [hunk 2]
```diff
@@ -335,7 +353,7 @@ def add_tour_leg(self, agent, k, o_zone, o_loc, d_zone, d_loc, start_tm, end_tm)
 
         return end_tm
 
-    def add_return_origin(self, agent, k, o_loc, d_zone, d_loc, end_tm, speed=50000/3600):
```
Comment (divyasharma-arup) [inline pam/samplers/tour.py:338]: this `speed=50000/3600` wasn't being used anywhere.

File: pam/samplers/tour.py [hunk 3]
```diff
@@ -434,6 +453,13 @@ def __init__(self, trips, zone, o_dist, d_dist, o_activity, d_activity, o_freq,
         self.od_density.rename(columns={'key_0':zone.index.name}, inplace=True)
         self.od_density.set_index(zone.index.name, inplace=True)
 
+        # Add in features for analysis
+        self.od_density = self.od_density.fillna(0)
```
Comment (divyasharma-arup) [inline pam/samplers/tour.py:457]: this all contributes to the `plot_density_difference function`

File: pam/samplers/tour.py [hunk 4]
```diff
@@ -457,6 +483,8 @@ def plot_validate_spatial_density(self, title_1, title_2, density_metric, densit
         cax = fig.add_axes([1,0.2,0.03,0.6]) 
         plt.colorbar(im, cax=cax)
 
+        return fig
```
Comment (divyasharma-arup) [inline pam/samplers/tour.py:486]: returning fig to enable testing

File: pam/samplers/tour.py (hunk 1 above)
Comment (fredshone) [reply pam/samplers/tour.py:154]: I'm curious about the warnings "module", what do the outputs look like? We use the logging module [else where](https://github.com/arup-group/pam/blob/893657b7617553b8756f562f671f636b12fe8972/pam/cli.py#L19). But this seems more light-weight and simpler to use.

**Review by fredshone — COMMENTED** [review]
Great - nice job adding/improving some doc-strings. No objections from me if you want to merge

File: pam/samplers/tour.py (hunk 1 above)
Comment (divyasharma-arup) [reply pam/samplers/tour.py:154]: It'll output as the below:

> /Users/divya.sharma/Documents/repos/pam/pam/samplers/tour.py:50: UserWarning:
> 
> Your density gdf has infinite values

I essentially googled 'pytest warnings' to try and understand why raise UserWarning wasn't in fact raising warnings, and it led me to the above solution ([link](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html)).

**Comment by divyasharma-arup** [comment]
> 



> Great - nice job adding/improving some doc-strings. No objections from me if you want to merge

Thanks Fred, once you hit Approve I can merge.

**Review by fredshone — APPROVED** [review]

**Pull request merged**